src/Http/Middleware/CsrfProtectionMiddleware.php

```diff
@@ -193,5 +199,19 @@ protected function _validateToken(ServerRequest $request)
         if ($post !== $cookie && $header !== $cookie) {
             throw new InvalidCsrfTokenException(__d('cake', 'CSRF token mismatch.'));
         }
+
+        try {
+            $decrypted = json_decode(Security::decrypt(hex2bin($cookie), Security::getSalt()), true);
+            if (!$decrypted) {
+                throw new InvalidCsrfTokenException(__d('cake', 'Invalid CSRF token'));
+            }
+            $expiry = new Time('@' . $decrypted['expiry']);
+            $now = new Time();
+            if ($now > $expiry) {
+                throw new InvalidCsrfTokenException(__d('cake', 'Expired CSRF token'));
+            }
+        } catch (\InvalidArgumentException $e) {
+            throw new InvalidCsrfTokenException(__d('cake', 'Invalid CSRF token'));
+        }
```

Comment:
Should we consider extracting the token creation/validation to a helper object that can be used by the middleware and component? It could be an internal class so we don't expose more api surface.